core/src/main/java/org/mule/runtime/core/internal/routing/ForeachRouter.java

```diff
@@ -0,0 +1,332 @@
+          }
+
+          TypedValue currentValue = setCurrentValue(batchSize, iterator);
+          return createTypedValuePartToProcess(owner, event, foreachContext, currentValue);
+
+        })
+        .transform(innerPub -> applyWithChildContext(innerPub, nestedChain, of(owner.getLocation())))
+        .doOnNext(evt -> {
+          try {
+            ForeachContext foreachContext = foreachContextResolver.getCurrentContextFromEvent(evt).get(evt.getContext().getId());
+
+            if (foreachContext.getOnComplete().isPresent()) {
+              foreachContext.getOnComplete().get().run();
+            }
+            // Check if I have more to iterate:
+            if (foreachContext.getIterator().hasNext()) {
+              // YES - Inject again inside innerFlux. The Iterator automatically keeps track of the following elements
+              innerRecorder.next(evt);
+            } else {
+              // NO - Propagate the first inside event down to downstreamFlux
+              downstreamRecorder.next(right(evt));
+              completeRouterIfNecessary();
+            }
+          } catch (Exception e) {
+            // Delete foreach context
+            this.eventWithCurrentContextDeleted(evt);
+            downstreamRecorder.next(left(new MessagingException(evt, e, owner)));
+            completeRouterIfNecessary();
+          }
+        }).onErrorContinue((e, o) -> {
+          this.eventWithCurrentContextDeleted(((MessagingException) e).getEvent());
+          downstreamRecorder.next(left(e));
+          completeRouterIfNecessary();
+        });
+
+    downstreamFlux = Flux.<Either<Throwable, CoreEvent>>create(sink -> {
+      downstreamRecorder.accept(sink);
+      // This will always run after the `downstreamCtxReference` is set
+      subscribeUpstreamChains(downstreamCtxReference.get());
+    })
+        .doOnNext(event -> inflightEvents.decrementAndGet())
+        .map(either -> {
+          if (either.isLeft()) {
+            throw propagate(either.getLeft());
+          } else {
+            // Create response and restore variables
+            return createResponseEvent(either.getRight());
+          }
+        });
+  }
+
+  /**
+   * If there are no events in-flight and the upstream publisher has received a completion signal, complete downstream publishers.
+   */
+  private void completeRouterIfNecessary() {
+    if (completeDeferred.get() && inflightEvents.get() == 0) {
+      completeRouter();
+    }
+  }
+
+
+  private void completeRouter() {
+    innerRecorder.complete();
+    downstreamRecorder.complete();
+  }
+
+
+  private TypedValue setCurrentValue(int batchSize, Iterator<TypedValue<?>> iterator) {
+    TypedValue currentValue;
+    if (batchSize > 1) {
+      int counter = 0;
+      List list = new ArrayList<>();
+      while (iterator.hasNext() && counter < batchSize) {
+        list.add(iterator.next());
+        counter++;
+      }
+      currentValue = new TypedValue<>(list, fromObject(list));
```

Comment:
can the result from `fromObject(list)` be cached instead of calculated for every batch?